packages/marshal/marshal.js

```diff
@@ -61,8 +74,27 @@ function pureCopy(val, already = new WeakMap()) {
       return harden(copy);
     }
 
+    case 'copyError': {
+      const err = /** @type {Error} */ (val);
+
+      if (already.has(err)) {
+        return already.get(err);
+      }
+
+      const copy = Error(err.message);
```

Comment:
That's not a valid way to copy an error. See how we use `getErrorContructor` in unserialization.

Here are interesting invariants to use as joint correctness criteria for `pureCopy`, `marshal`, and `sameStructure`:
For all `x`
   * `pureCopy(x)` succeeds ***iff*** marshal(x) succeeds with no slots.
   * Given `y = pureCopy(x)`, then `sameStructure(x, y)`
   * Given `z = unmarshal(marshal(x))`, then `sameStructure(x, z)`
   * sameStructure is an equivalence class (transitive, reflexive, commutative) so `sameStructure(y, z)` as well
   * Unlike `y` and `z`, `x` may be or contain proxies. This cannot cause any violation of the above points under success, but it can cause some cases involving `x` to throw, not terminate, or cause side effects on the side. That's in fact the motivation of making the copy --- to insulate us from those forms of proxy-based interference.

Anything about either `marshal`, `pureCopy`, or `sameStructure` we are uncertain about, we should check against these invariants.

The error issue is the only one that i know of.

Reply:
Fixed!